src/hidimstat/base_perturbation.py

```diff
@@ -0,0 +1,132 @@
+        loss : callable, default=root_mean_squared_error
+            The loss function to use when comparing the perturbed model to the full
+            model.
+        n_permutations : int, default=50
+            Only for PermutationImportance or CPI. The number of permutations to
+            perform.
+        method : str, default="predict"
+            The method to use for the prediction. This determines the predictions passed
+            to the loss function.
+        n_jobs : int, default=1
+            The number of jobs to run in parallel. Parallelization is done over the
+            variables or groups of variables.
+
+        """
+        check_is_fitted(estimator)
+        self.estimator = estimator
+        self.loss = loss
+        _check_vim_predict_method(method)
+        self.method = method
+        self.n_jobs = n_jobs
+        self.n_permutations = n_permutations
+        self.n_groups = None
+
+    def fit(self, X, y, groups=None):
+        if groups is None:
+            self.n_groups = X.shape[1]
+            self.groups = {j: [j] for j in range(self.n_groups)}
+            self._groups_ids = np.array(list(self.groups.values()), dtype=int)
+        else:
+            self.n_groups = len(groups)
+            self.groups = groups
+            if isinstance(X, pd.DataFrame):
+                self._groups_ids = []
+                for group_key in self.groups.keys():
+                    self._groups_ids.append(
+                        [
+                            i
+                            for i, col in enumerate(X.columns)
+                            if col in self.groups[group_key]
+                        ]
+                    )
+            else:
+                self._groups_ids = np.array(list(self.groups.values()), dtype=int)
+
+    def predict(self, X):
+        self._check_fit()
+        X_ = np.asarray(X)
+
+        # Parallelize the computation of the importance scores for each group
+        out_list = Parallel(n_jobs=self.n_jobs)(
+            delayed(self._joblib_predict_one_group)(X_, group_id, group_key)
+            for group_id, group_key in enumerate(self.groups.keys())
+        )
+        return np.stack(out_list, axis=0)
+
+    def score(self, X, y):
+        self._check_fit()
+
+        out_dict = dict()
+
+        y_pred = getattr(self.estimator, self.method)(X)
+        loss_reference = self.loss(y, y_pred)
+        out_dict["loss_reference"] = loss_reference
+
+        y_pred = self.predict(X)
+        out_dict["loss"] = dict()
+        for j, y_pred_j in enumerate(y_pred):
+            list_loss = []
+            for y_pred_perm in y_pred_j:
+                list_loss.append(self.loss(y, y_pred_perm))
+            out_dict["loss"][j] = np.array(list_loss)
+
+        out_dict["importance"] = np.array(
+            [
+                np.mean(out_dict["loss"][j]) - loss_reference
+                for j in range(self.n_groups)
+            ]
+        )
+        return out_dict
```

Comment:
Wondering how (non-)standard it is to return a dictionary from a `score` method ?

Reply:
Good question; it is done in some of the cross_val functions from sklearn (returning time, scores for different metrics/splits, fitted estimators ...). 
I think it is not a necessity here. We might as well return only the importance. 
The other values are : 
 - the `loss_reference` (of the non-perturbed model), which can be computed outside in a purely "predictive" way. 
 - the `loss` corresponding to the loss of the perturbed model. Here, removing this output would remove the possibility of accessing the variability over permutations as the importance is computing after averaging over permutatoins.